File: packages/virage-cli/src/cli/update.ts
import { checkbox } from "@inquirer/prompts";
import { mkdir, readFile, writeFile } from "fs/promises";
import { existsSync } from "fs";
import { join } from "path";
import { execFile } from "child_process";
import { promisify } from "util";
import { createOut } from "../output.js";
import { discoverAgentPlugins, runAgentPlugin } from "./agent-plugin.js";
import { resolveSkillsPackagePath, syncSkills } from "./skills.js";
import {
  detectPackageManager,
  buildInstallCommand,
  buildGlobalInstallCommand,
  getLocalPluginDir,
  getGlobalPluginDir,
  buildPluginPrefixInstallCommand,
  runInstall,
} from "./pkg-manager.js";

const execFileAsync = promisify(execFile);
const isWin = process.platform === "win32";
const npmBin = isWin ? "npm.cmd" : "npm";

interface PackageJson {
  name?: string;
  version?: string;
  dependencies?: Record<string, string>;
  devDependencies?: Record<string, string>;
  "rag-plugin"?: unknown;
  "virage-agent"?: unknown;
}

interface VirageConfig {
  providers?: {
    embedder?: { package?: string };
    vectorStore?: { package?: string };
    reranker?: { package?: string };
    source?: { package?: string };
  };
  search?: { reranker?: { package?: string } };
  agents?: Array<{ package: string } | string>;
  fileSets?: Array<{ chunkers?: Array<{ package?: string }> }>;
}

const AGENT_PACKAGES: Record<string, string> = {
  "claude-code": "@vivantel/virage-agent-claude",
  copilot: "@vivantel/virage-agent-copilot",
  codex: "@vivantel/virage-agent-codex",
  antigravity: "@vivantel/virage-agent-antigravity",
};

async function readProjectPackageJson(
  cwd: string,
): Promise<PackageJson | null> {
  const pkgPath = join(cwd, "package.json");
  if (!existsSync(pkgPath)) return null;
  try {
    const raw = await readFile(pkgPath, "utf-8");
    return JSON.parse(raw) as PackageJson;
  } catch {
    return null;
  }
}

async function readNodeModulePackageJson(
  cwd: string,
  packageName: string,
): Promise<PackageJson | null> {
  const pkgPath = join(cwd, "node_modules", packageName, "package.json");
  if (!existsSync(pkgPath)) return null;
  try {
    const raw = await readFile(pkgPath, "utf-8");
    return JSON.parse(raw) as PackageJson;
  } catch {
    return null;
  }
}

async function readVirageConfig(
  configPath: string,
): Promise<VirageConfig | null> {
  if (!existsSync(configPath)) return null;
  try {
    const raw = await readFile(configPath, "utf-8");
    return JSON.parse(raw) as VirageConfig;
  } catch {
    return null;
  }
}

async function discoverViragePackages(
  cwd: string,
  configPath: string,
  isGlobal: boolean,
): Promise<string[]> {
  const candidates = new Set<string>();

  // Always include virage-core and virage-skills — required by every project
  candidates.add("@vivantel/virage-core");
  candidates.add("@vivantel/virage-skills");

  // Config-first: read virage.config.json to discover required packages
  const virageConfig = await readVirageConfig(configPath);
  if (virageConfig) {
    if (virageConfig.providers?.embedder?.package) {
      candidates.add(virageConfig.providers.embedder.package);
    }
    if (virageConfig.providers?.vectorStore?.package) {
      candidates.add(virageConfig.providers.vectorStore.package);
    }
    if (virageConfig.providers?.reranker?.package) {
      candidates.add(virageConfig.providers.reranker.package);
    }
    if (virageConfig.providers?.source?.package) {
      candidates.add(virageConfig.providers.source.package);
    }
    if (virageConfig.search?.reranker?.package) {
      candidates.add(virageConfig.search.reranker.package);
    }
    for (const agent of virageConfig.agents ?? []) {
      if (typeof agent === "string") {
        const pkg = AGENT_PACKAGES[agent];
        if (pkg) candidates.add(pkg);
      } else if (agent.package) {
        candidates.add(agent.package);
      }
    }
    for (const fileSet of virageConfig.fileSets ?? []) {
      for (const chunker of fileSet.chunkers ?? []) {
        if (chunker.package) candidates.add(chunker.package);
      }
    }
  }

  // Supplement: if package.json exists, also scan it for any @vivantel/* or
  // rag-plugin/virage-agent packages not already covered by config
  if (!isGlobal) {
    const projectPkg = await readProjectPackageJson(cwd);
    if (projectPkg) {
      const allDeps = {
        ...projectPkg.dependencies,
        ...projectPkg.devDependencies,
      };
      for (const pkgName of Object.keys(allDeps)) {
        if (pkgName.startsWith("@vivantel/")) {
          candidates.add(pkgName);
          continue;
        }
        const depPkg = await readNodeModulePackageJson(cwd, pkgName);
        if (depPkg && (depPkg["rag-plugin"] || depPkg["virage-agent"])) {
          candidates.add(pkgName);
        }
      }
    }
  }

  return Array.from(candidates);
}

async function getLatestVersion(packageName: string): Promise<string | null> {
  try {
    const { stdout } = await execFileAsync(
      npmBin,
      ["view", packageName, "version", "--json", "--prefer-online"],
      { timeout: 10_000, shell: isWin },
    );
    return (JSON.parse(stdout.trim()) as string).trim();
  } catch {
    return null;
  }
}

async function readPluginDirPackageJson(
  pluginDir: string,
  packageName: string,
): Promise<PackageJson | null> {
  const pkgPath = join(pluginDir, "node_modules", packageName, "package.json");
  if (!existsSync(pkgPath)) return null;
  try {
    const raw = await readFile(pkgPath, "utf-8");
    return JSON.parse(raw) as PackageJson;
  } catch {
    return null;
  }
}

async function getCurrentVersion(
  cwd: string,
  packageName: string,
): Promise<string | null> {
  // Check local plugin dir first (highest priority)
  const localPluginPkg = await readPluginDirPackageJson(
    getLocalPluginDir(cwd),
    packageName,
  );
  if (localPluginPkg?.version) return localPluginPkg.version;

  // Check global plugin dir
  const globalPluginPkg = await readPluginDirPackageJson(
    getGlobalPluginDir(),
    packageName,
  );
  if (globalPluginPkg?.version) return globalPluginPkg.version;

  // Try local node_modules (backwards compat)
  const pkg = await readNodeModulePackageJson(cwd, packageName);
  if (pkg?.version) return pkg.version;

  // Fallback: query npm for the installed version (works for global installs)
  try {
    const { stdout } = await execFileAsync(
      npmBin,
      ["list", packageName, "--json", "--depth=0"],
      { timeout: 10_000, shell: isWin },
    );
    const parsed = JSON.parse(stdout) as {
      dependencies?: Record<string, { version?: string }>;
    };
    const version = parsed.dependencies?.[packageName]?.version;
    return version ?? null;
  } catch {
    return null;
  }
}

async function getPackageInstallLocation(
  cwd: string,
  packageName: string,
): Promise<"local-plugin" | "global-plugin" | "node_modules" | "global-npm"> {
  if (
    (await readPluginDirPackageJson(getLocalPluginDir(cwd), packageName))
      ?.version
  )
    return "local-plugin";
  if (
    (await readPluginDirPackageJson(getGlobalPluginDir(), packageName))?.version
  )
    return "global-plugin";
  if ((await readNodeModulePackageJson(cwd, packageName))?.version)
    return "node_modules";
  return "global-npm";
}

interface PackageStatus {
  name: string;
  current: string;
  latest: string;
  outdated: boolean;
  location: "local-plugin" | "global-plugin" | "node_modules" | "global-npm";
}

export async function runUpdate(
  configPath: string,
  verbosity = 0,
  opts: { force?: boolean; yes?: boolean } = {},
): Promise<void> {
  const out = createOut(verbosity);
  const cwd = process.cwd();

  const hasPackageJson = existsSync(join(cwd, "package.json"));
  const hasVirageConfig = existsSync(configPath);

  if (!hasPackageJson && !hasVirageConfig) {
    out.warn(`No config file found at ${configPath}.`);
    out.dim("Run `virage init` first to configure your project.");
    return;
  }

  const isGlobal = !hasPackageJson;
  if (isGlobal) {
    out.dim(
      "(No package.json found — treating as standalone project, will install globally.)",
    );
  }

  out.section("Virage Update");
  out.dim("Discovering virage ecosystem packages...");

  const packageNames = await discoverViragePackages(cwd, configPath, isGlobal);

  // Fetch current and latest versions in parallel
  const statuses = await Promise.all(
    packageNames.map(async (name): Promise<PackageStatus> => {
      const [current, latest, location] = await Promise.all([
        getCurrentVersion(cwd, name),
        getLatestVersion(name),
        getPackageInstallLocation(cwd, name),
      ]);
      const cur = current ?? "unknown";
      const lat = latest ?? "unknown";
      return {
        name,
        current: cur,
        latest: lat,
        outdated: cur !== "unknown" && lat !== "unknown" && cur !== lat,
        location,
      };
    }),
  );

  const outdated = statuses.filter((s) => s.outdated);
  const unknownLatest = statuses.filter((s) => s.latest === "unknown");

  if (outdated.length === 0) {
    out.success("All virage packages are up to date.");
  } else {
    out.info(`\nFound ${outdated.length} outdated package(s):`);
    for (const s of outdated) {
      out.dim(`  ${s.name}: ${s.current} → ${s.latest}`);
    }
  }
  if (unknownLatest.length > 0) {
    out.warn(
      `Could not fetch latest version for ${unknownLatest.length} package(s) (registry unreachable?):`,
    );
    for (const s of unknownLatest) {
      out.dim(`  ${s.name}: ${s.current}`);
    }
  }

  if (statuses.length === 0) {
    out.warn("No virage packages to update.");
    return;
  }

  // Let user pick which packages to update (skipped with --yes)
  let toUpdate: string[];
  if (opts.yes) {
    toUpdate = statuses.map((s) => s.name);
    out.dim("Updating all packages (--yes).");
  } else {
    toUpdate = await checkbox({
      message: "Select packages to update:",
      choices: statuses.map((s) => ({
        name: s.outdated
          ? `${s.name}  (${s.current} → ${s.latest})`
          : s.latest === "unknown"
            ? `${s.name}  (${s.current}, latest unknown)`
            : `${s.name}  (${s.current}, up to date)`,
        value: s.name,
        checked: true,
      })),
    });
  }

  if (toUpdate.length > 0) {
    const pm = await detectPackageManager(cwd);
    const forceArgs = opts.force ? ["--force"] : [];

    // Split packages by install location
    const selectedStatuses = statuses.filter((s) => toUpdate.includes(s.name));
    const localPluginPkgs = selectedStatuses
      .filter((s) => s.location === "local-plugin")
      .map((s) => `${s.name}@latest`);
    const globalPluginPkgs = selectedStatuses
      .filter((s) => s.location === "global-plugin")
      .map((s) => `${s.name}@latest`);
    const nodeModulesPkgs = selectedStatuses
      .filter(
        (s) => s.location === "node_modules" || s.location === "global-npm",
      )
      .map((s) => `${s.name}@latest`);

    let updateFailed = false;

    if (localPluginPkgs.length > 0) {
      const localPluginDir = getLocalPluginDir(cwd);
      await mkdir(localPluginDir, { recursive: true });
      const { cmd, args } = buildPluginPrefixInstallCommand(
        localPluginPkgs,
        localPluginDir,
      );
      const fullArgs = [...args, ...forceArgs];
      out.dim(`\nRunning: ${cmd} ${fullArgs.join(" ")}`);
      try {
        await runInstall(cmd, fullArgs);
      } catch {
        out.error("Local plugin update failed. Try running manually:");
        out.dim(`  ${cmd} ${fullArgs.join(" ")}`);
        updateFailed = true;
      }
    }

    if (globalPluginPkgs.length > 0) {
      const globalPluginDir = getGlobalPluginDir();
      await mkdir(globalPluginDir, { recursive: true });
      const { cmd, args } = buildPluginPrefixInstallCommand(
        globalPluginPkgs,
        globalPluginDir,
      );
      const fullArgs = [...args, ...forceArgs];
      out.dim(`\nRunning: ${cmd} ${fullArgs.join(" ")}`);
      try {
        await runInstall(cmd, fullArgs);
      } catch {
        out.error("Global plugin update failed. Try running manually:");
        out.dim(`  ${cmd} ${fullArgs.join(" ")}`);
        updateFailed = true;
      }
    }

    if (nodeModulesPkgs.length > 0) {
      const { cmd, args } = isGlobal
        ? buildGlobalInstallCommand(pm, nodeModulesPkgs)
        : buildInstallCommand(pm, nodeModulesPkgs);
      const fullArgs = [...args, ...forceArgs];
      out.dim(`\nRunning: ${cmd} ${fullArgs.join(" ")}`);
      try {
        await runInstall(cmd, fullArgs);
      } catch {
        out.error("Update failed. Try running manually:");
        out.dim(`  ${cmd} ${fullArgs.join(" ")}`);
        updateFailed = true;
      }
    }

    if (!updateFailed) {
      out.success("Packages updated successfully.");

      // Rewrite pluginVersions in virage config for plugin-dir packages
      if (existsSync(configPath)) {
        try {
          const raw = await readFile(configPath, "utf-8");
          const cfg = JSON.parse(raw) as Record<string, unknown>;
          const existingVersions =
            (cfg.pluginVersions as Record<string, string> | undefined) ?? {};
          for (const s of selectedStatuses) {
            if (
              s.location === "local-plugin" ||
              s.location === "global-plugin" ||
              s.name in existingVersions
            ) {
              existingVersions[s.name] = `~${s.latest}`;
            }
          }
          cfg.pluginVersions = existingVersions;
          await writeFile(configPath, JSON.stringify(cfg, null, 2) + "\n");
        } catch {
          // Non-fatal: config update failed but packages were installed
        }
      }
    }
  }

  // Always re-run agent plugin configure() to sync plugin-config files
  const agentPlugins = await discoverAgentPlugins(cwd);
  if (agentPlugins.length > 0) {
    out.dim("\nSyncing agent plugin configs...");
    for (const plugin of agentPlugins) {
      try {
        const result = await runAgentPlugin(plugin, cwd);
        const msg = result.hooksWritten
          ? "config updated"
          : "already up to date";
        const mcpMsg =
          result.mcpRegistered === true
            ? "; MCP server registered"
            : result.mcpRegistered === false
              ? "; MCP server already registered"
              : "";
        out.dim(`  ${plugin.label}: ${msg}${mcpMsg}`);
      } catch (err) {
        out.warn(
          `  ${plugin.label} failed: ${err instanceof Error ? err.message : String(err)}`,
        );
      }
    }
  }

  // Self-update @vivantel/virage-cli in global npm if globally installed
  try {
    const { stdout: listOut } = await execFileAsync(
      npmBin,
      ["list", "-g", "--depth=0", "--json"],
      { timeout: 10_000, shell: isWin },
    );
    const listed = JSON.parse(listOut) as {
      dependencies?: Record<string, { version?: string }>;
    };
    const cliEntry = listed.dependencies?.["@vivantel/virage-cli"];
    if (cliEntry) {
      const currentCliVersion = cliEntry.version ?? "unknown";
      const latestCliVersion = await getLatestVersion("@vivantel/virage-cli");
      if (latestCliVersion && currentCliVersion !== latestCliVersion) {
        out.dim(
          `\nUpdating @vivantel/virage-cli: ${currentCliVersion} → ${latestCliVersion}`,
        );
        await runInstall("npm", [
          "install",
          "-g",
          "@vivantel/virage-cli@latest",
        ]);
        out.dim("  CLI updated. New version active on next invocation.");
      } else {
        out.dim("\n@vivantel/virage-cli is up to date.");
      }
    }
  } catch {
    // Non-fatal: CLI self-update is best-effort
  }

  // Auto-sync skills if available — no interactive prompt needed
  const skillsPkgPath = resolveSkillsPackagePath();
  if (skillsPkgPath !== null) {
    out.dim("\nSyncing Virage AI agent skills...");
    try {
      const result = await syncSkills(skillsPkgPath, cwd);
      if (result.created.length > 0)
        out.dim(`  Skills installed: ${result.created.length} new`);
      if (result.updated.length > 0)
        out.dim(`  Skills updated: ${result.updated.length}`);
      if (result.deleted.length > 0)
        out.dim(`  Skills removed: ${result.deleted.length}`);
      if (
        result.created.length === 0 &&
        result.updated.length === 0 &&
        result.deleted.length === 0
      )
        out.dim("  Already up to date.");
    } catch (err) {
      out.warn(
        `  Skills sync failed: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
  }

  out.success("Done.");
}

File: crates/virage-engine/src/cli/update.rs
use std::path::Path;

use crate::config::find_config;
use crate::output::{Out, OutputFormat};

use super::util::{multiselect_step, spinner};

/// Returns the correct npm binary name for the current OS.
/// On Windows, `npm` is a batch script (`npm.cmd`), not a native executable.
fn npm_bin() -> &'static str {
    if cfg!(windows) {
        "npm.cmd"
    } else {
        "npm"
    }
}

struct PackageStatus {
    name: String,
    current: String,
    latest: String,
    outdated: bool,
}

/// Queries `npm view <pkg> version --json` and returns the latest published version.
fn get_npm_latest(npm: &str, pkg: &str) -> Option<String> {
    let output = std::process::Command::new(npm)
        .args(["view", pkg, "version", "--json", "--prefer-online"])
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    let s = String::from_utf8_lossy(&output.stdout).trim().to_string();
    serde_json::from_str::<String>(&s).ok()
}

/// Returns the currently installed version of a package, checking local node_modules
/// then falling back to `npm list --global`.
fn get_npm_current(npm: &str, pkg: &str, cwd: &Path) -> Option<String> {
    let local_pkg = cwd.join("node_modules").join(pkg).join("package.json");
    if let Ok(raw) = std::fs::read_to_string(local_pkg) {
        if let Ok(v) = serde_json::from_str::<serde_json::Value>(&raw) {
            if let Some(ver) = v.get("version").and_then(|v| v.as_str()) {
                return Some(ver.to_string());
            }
        }
    }
    let out = std::process::Command::new(npm)
        .args(["list", pkg, "--global", "--json", "--depth=0"])
        .output()
        .ok()?;
    let s = String::from_utf8_lossy(&out.stdout);
    let v: serde_json::Value = serde_json::from_str(&s).ok()?;
    v.get("dependencies")
        .and_then(|d| d.get(pkg))
        .and_then(|p| p.get("version"))
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
}

/// Discovers all @vivantel/* packages referenced in virage.config.json and package.json.
fn discover_virage_packages(cwd: &Path, config_path: &str) -> Vec<String> {
    let mut packages = std::collections::BTreeSet::new();

    if let Ok(raw) = std::fs::read_to_string(config_path) {
        if let Ok(v) = serde_json::from_str::<serde_json::Value>(&raw) {
            if let Some(providers) = v.get("providers").and_then(|p| p.as_object()) {
                for provider in providers.values() {
                    if let Some(pkg) = provider.get("package").and_then(|p| p.as_str()) {
                        packages.insert(pkg.to_string());
                    }
                }
            }
            if let Some(agents) = v.get("agents").and_then(|a| a.as_array()) {
                for agent in agents {
                    if let Some(pkg) = agent.get("package").and_then(|p| p.as_str()) {
                        packages.insert(pkg.to_string());
                    }
                }
            }
            if let Some(file_sets) = v.get("fileSets").and_then(|f| f.as_array()) {
                for fs in file_sets {
                    if let Some(chunkers) = fs.get("chunkers").and_then(|c| c.as_array()) {
                        for chunker in chunkers {
                            if let Some(pkg) = chunker.get("package").and_then(|p| p.as_str()) {
                                packages.insert(pkg.to_string());
                            }
                        }
                    }
                }
            }
        }
    }

    let pkg_json = cwd.join("package.json");
    if let Ok(raw) = std::fs::read_to_string(pkg_json) {
        if let Ok(v) = serde_json::from_str::<serde_json::Value>(&raw) {
            for key in ["dependencies", "devDependencies"] {
                if let Some(deps) = v.get(key).and_then(|d| d.as_object()) {
                    for name in deps.keys() {
                        if name.starts_with("@vivantel/") {
                            packages.insert(name.clone());
                        }
                    }
                }
            }
        }
    }

    packages.into_iter().collect()
}

pub fn cmd_update(verbose: u8, format: OutputFormat) -> anyhow::Result<()> {
    let out = Out::new(verbose, format);
    use console::style;

    let npm = npm_bin();
    let cwd = std::env::current_dir()?;

    out.section("Virage Update");

    // ── 1. Discover packages ──────────────────────────────────────────────────
    let config_path = find_config().unwrap_or_else(|| "virage.config.json".into());
    let packages = discover_virage_packages(&cwd, &config_path);

    if packages.is_empty() {
        out.warn("No @vivantel/* packages found.");
        return Ok(());
    }

    // ── 2. Check versions ─────────────────────────────────────────────────────
    let pb = spinner("Checking versions...");
    let mut statuses: Vec<PackageStatus> = Vec::new();
    for pkg in &packages {
        pb.set_message(format!("Checking {pkg}..."));
        let current = get_npm_current(npm, pkg, &cwd).unwrap_or_else(|| "not installed".into());
        let latest = get_npm_latest(npm, pkg).unwrap_or_else(|| "unknown".into());
        let outdated = latest != "unknown" && current != "not installed" && current != latest;
        statuses.push(PackageStatus {
            name: pkg.clone(),
            current,
            latest,
            outdated,
        });
    }
    pb.finish_and_clear();

    // ── 3. Display status table ───────────────────────────────────────────────
    println!();
    for s in &statuses {
        let cur_styled = if s.outdated {
            style(&s.current).yellow().to_string()
        } else if s.current == "not installed" {
            style(&s.current).dim().to_string()
        } else {
            style(&s.current).green().to_string()
        };
        let lat_styled = if s.outdated {
            style(&s.latest).green().to_string()
        } else {
            style(&s.latest).dim().to_string()
        };
        println!(
            "  {:45}  {}  →  {}",
            style(&s.name).dim(),
            cur_styled,
            lat_styled
        );
    }
    println!();

    // ── 4. Interactive selection ──────────────────────────────────────────────
    let labels: Vec<String> = statuses
        .iter()
        .map(|s| {
            if s.outdated {
                format!("{} ({} → {})", s.name, s.current, s.latest)
            } else {
                format!("{} ({})", s.name, s.current)
            }
        })
        .collect();

    let default_selected: Vec<usize> = statuses
        .iter()
        .enumerate()
        .filter_map(|(i, s)| s.outdated.then_some(i))
        .collect();

    let selected = match multiselect_step("Packages to update", &labels, &default_selected)? {
        None => {
            out.info("Cancelled.");
            return Ok(());
        }
        Some(s) => s,
    };

    if selected.is_empty() {
        out.info("Nothing selected.");
        return Ok(());
    }

    // ── 5. Install selected packages — I1: single batched npm install -g call ──
    let to_install: Vec<&str> = selected
        .iter()
        .map(|&i| statuses[i].name.as_str())
        .collect();
    out.info(&format!("Installing {} package(s)...", to_install.len()));

    let pkg_args: Vec<String> = to_install.iter().map(|p| format!("{p}@latest")).collect();
    out.dim(&format!("  npm install -g {}", pkg_args.join(" ")));
    let status = std::process::Command::new(npm)
        .arg("install")
        .arg("-g")
        .args(&pkg_args)
        .status();
    match status {
        Ok(s) if s.success() => out.success(&format!("{} package(s) updated", to_install.len())),
        Ok(s) => out.warn(&format!("npm exited with status {s}")),
        Err(e) => out.error(&format!("npm install failed: {e}")),
    }

    // ── 6. Self-update (virage CLI binary) ────────────────────────────────────
    out.dim("Checking virage binary...");
    let self_current =
        get_npm_current(npm, "@vivantel/virage", &cwd).unwrap_or_else(|| "unknown".into());
    let self_latest = get_npm_latest(npm, "@vivantel/virage").unwrap_or_else(|| "unknown".into());

    if self_latest != "unknown" && self_current != self_latest {
        out.info(&format!(
            "Updating virage binary {self_current} → {self_latest}..."
        ));
        let status = std::process::Command::new(npm)
            .args(["install", "-g", "@vivantel/virage@latest"])
            .status();
        match status {
            Ok(s) if s.success() => out.success("virage binary updated."),
            Ok(s) => out.warn(&format!("virage binary update exited {s}")),
            Err(e) => out.error(&format!("virage binary update failed: {e}")),
        }
    } else {
        out.dim("virage binary is up to date.");
    }

    out.success("Update complete.");
    Ok(())
}
